content/blog/2023-12-11-circ-bitwidth-adjustment

```diff
@@ -0,0 +1,115 @@
+
+## Design
+Our optimization mainly focuses on addition and multiplication, which would typically increase the bitwidth of resulting element by 1 or double the bitwidth, respectively. We classify elements into two types
+- Elements that could be above the range and required to be returned within range: this includes program output, elements involved in bitwise operations and comparisons, operations like division or remainder, elements involved in memory operations, and so on.
+- Elements that could be above the range but are not required to be within range: This includes intermediate elements that undergoes multiplication and addition operations and does not meet any of the criterions above.
+
+The current system represents basic elements as bitvectors, and is designed to adjust their bitwidth after each arithmetic operation. Based on the classification of the elements above, our projects reduces the number of constraints in the system by greedily deciding when to adjust bitwidth for elements falls under the second category while ensuring no overflows can happen in later operations involving the elements. Or in other words, adjustments are only introduced if the next operation is going to result in an overflow.
+
+Take this short program as an example:
+```
+def main(u32 a, u32 b) -> u32:
+    u32 c = a + b
+    u32 d = a + c
+    return c * d
+```
+
+The old compiler would adjust immediately at `a+b`, `a+c` and `c * d`. By applying our optimization, the system would treat c and d as intermediate elements and only adjust at c*d, since it is producing the final program output. 
+Without our optimization, the cost spent on adjusting the bitwidth will be: 33 constraint to adjust `a + b` and `a + c` respectively, and 64 to adjust `c * d`, which result in 130 constraints in total.
+In contrast, after our optimization, the cost spent on adjusting `a + b` and `a + c` would be 0, and 66 to adjust `c * d`, which result in 66 constraints in total. 
+
+## Implementation
+1. We create a new type of operations: `add_unadjust` and `mul_unadjust` for addition and multiplication to notify the constraint generation system not to adjust the bitwidth.
+2. We implemented two passes to rewrite the operations. First pass is to classify elements in the system. We implemented an initial pass to scan the entire program and store elements requiring adjustments in a set S. Whenever we encounter an operation that definitely requires adjustment, we add the elements used in the operation to S. Then we implemented a rewrite pass to rewrite operations done on bitvectors not included in D to our newly introduced unadjust operation. 
```

Comment:
But how does it decide whether the operations need adjustments? How does the classification actually work, in terms of a program analysis?